### PR TITLE
[B-01981] db encryption

### PR DESCRIPTION
**Pull request by neonphog**
- refactor of the ipc code to make requiring and actually using the passphrase possible
- addition of db_key encrypted file to manage the database key
- usage of the database key to encrypt the sqlcipher database

### CODE REVIEW

**Comment by Connoropolous** [comment]
@neonphog this sounds awesome, I'm just wondering, during what timeframe is a passphrase given... can you outline typical use?

**Comment by neonphog** [comment]
@Connoropolous 

> @neonphog this sounds awesome, I'm just wondering, during what moments are a passphrase set or used... can you outline typical use?

Systemically, this is still a WIP. The ultimate goal would be to have a dialog open up on a user's phone whenever their holoport needs to unlock the keystore. Right now, the passphrase is stored in holochain's config and auto-responds to lair's request... so obviously that's not providing any security against someone who has access to the drive contents.

**Comment by Connoropolous** [comment]
I see, i see. What defines when would the keystore need to be unlocked?

**Comment by neonphog** [comment]
@Connoropolous 

> I see, i see. What defines when would the keystore need to be unlocked?

Hopefully only once when Holochain first starts. Technically, every time it makes a new connection to Lair, but hopefully that connection will be stable enough it doesn't have to be re-established during runtime.

**Review by freesig — APPROVED** [review]
This looks good from a very high level but I really don't understand the code that well

File: crates/lair_keystore/src/lib.rs [hunk 1]
```diff
@@ -98,3 +101,176 @@ pub async fn execute_load_ed25519_keypair(
...
+        let mut dec = SecretStreamDecrypt::new(pre_key, self.header.clone())
+            .map_err(|e| {
+                LairError::from(format!("decrypt new fail: {:?}", e))
+            })?;
+
+        let db_key = sodoken::BufWriteSized::new_mem_locked()?;
+        dec.pull(
+            self.cipher.clone(),
+            <Option<sodoken::BufRead>>::None,
+            // erm... fix this in sodoken
```
[the remark this answers is not in the document]
Comment (neonphog) [reply crates/lair_keystore/src/lib.rs:214]: I'm missing some `From` implementations in sodoken that would make this more straight forward... aiight, good point, I'll go ahead and implement those before merging this : )

**Pull request merged**